### DISCUSSION TITLE
SSO authentication

### DISCUSSION THREAD

**Discussion by pengjiewang** [usnistgov/NEMO#240] [Q&A] [answered]
It seems that most institutions have moved to SAML SSO. Does NEMO support this? Any way to integrate SSO to NEMO?

**Comment by rptmat57** [comment, marked as answer]
NEMO itself does not deal with authentication.

The easiest way to use SSO with NEMO is by having SAML authentication done in Apache or Nginx and have them set a "remote user" header with the username of the authenticated user.

in NEMO's `settings.py`, you can then:

- add `"NEMO.middleware.HTTPHeaderAuthenticationMiddleware"` to middlewares
- set `AUTHENTICATION_BACKENDS = ["NEMO.views.authentication.RemoteUserAuthenticationBackend"]`.
- set `AUTHENTICATION_HEADER = "REMOTE_USER"` (to fetch the username from the `HTTP_REMOTE_USER` header)

**Reply by pengjiewang** [reply]
Thank you so mcuh! That sounds awesome!

**Reply by rptmat57** [reply]
if you are using shibboleth like many facilities/universities, we also provide a version of nginx that will work with that: 
https://hub.docker.com/r/nanofab/nginx-shibboleth

**Reply by Hydrozyk** [reply]
This is helpful, thanks. I wonder if you could post more info on how to get this working with nginx and shibboleth? That would be greatly appreciated! We are a lab with limited staff so not particular my area of expertise here.

**Reply by rptmat57** [reply]
I added a bit more information in the link on docker but the rest your IT should be able to help you out with

**Reply by Hydrozyk** [reply]
> I added a bit more information in the link on docker but the rest your IT should be able to help you out with

Thank you so much!

**Reply by Hydrozyk** [reply]
I am actually getting en error about zip when running this step:

```
docker exec -it nginx bash -c "zip -j /etc/shibboleth/shib-keys.zip /etc/shibboleth/sp-*"
bash: line 1: zip: command not found
```

But zip is installed on the system, so wonder if this is Docker image issue? 
```
zip -v
Copyright (c) 1990-2008 Info-ZIP - Type 'zip "-L"' for software license.
This is Zip 3.0 (July 5th 2008), by Info-ZIP.
```

**Comment by rptmat57** [comment]
Hi Paul, this instructions are referring to an old version of the image.
It will work if you remove the version number (it should use the latest): `docker run --name nginx -d nanofab/nginx-shibboleth`
you might have to remove the one you created previously `docker rm -f nginx`

**Reply by Hydrozyk** [reply]
I fixed the zip issue, thank you! 

I wonder if you could give any insight on how nanofab/nginx-shibboleth docker images works/interact with NEMO docker image? How to get them to talk to arch other? Sorry if asking here, let me know if there is a better method co communicate.

Appreciate for all your help!

**Reply by Hydrozyk** [reply]
Also, start up for nanofab/nginx-shibboleth docker failed with error:
docker logs nginx
NO DEBUG
2024/09/12 15:54:55 [emerg] 1#1: dlopen() "/etc/nginx/modules/ngx_http_shibboleth_module.so" failed (/etc/nginx/modules/ngx_http_shibboleth_module.so: cannot open shared object file: No such file or directory) in /etc/nginx/nginx.conf:1
nginx: [emerg] dlopen() "/etc/nginx/modules/ngx_http_shibboleth_module.so" failed (/etc/nginx/modules/ngx_http_shibboleth_module.so: cannot open shared object file: No such file or directory) in /etc/nginx/nginx.conf:1

**Reply by rptmat57** [reply]
The docker nginx container is the entry point and deals with the authentication and requests.
It acts as a proxy to nemo's container. NEMO's docker container is not directly exposed by docker, it is simply setup on the same docker network as nginx.
So essentially you have incoming requests coming from the outside to the ports that are exposed for Nginx. Then nginx uses the internal network to proxy the requests to NEMO (NEMO is not exposed directly to the outside).

The internal docker network is automatically created by docker if you are using docker compose. if you are not using docker compose, you have to create the network yourself and connect nginx and nemo to it.

between the nginx shibboleth configuration and the configuration for docker compose here: https://github.com/usnistgov/NEMO/wiki/Docker-Compose-installation you should be able to get everything going.

For the nginx error I would have to see your docker-compose.yml to check the volumes and everything.

**Reply by Hydrozyk** [reply]
Great, thanks. I got all this configured and set up. Got metadate from out IdP provider and set it up.  Shibbolthe works if I navigate to https://myhost.edu/Shibboleth.sso/Login and brings me to https://myhost.edu//login/?next=/ which it supped to login to Nemo but getting an error "There was an error pre-authenticating the user"

In settings.py I have set up as you mentioned above: 
```
 NEMO.middleware.HTTPHeaderAuthenticationMiddleware
AUTHENTICATION_BACKENDS = ["NEMO.views.authentication.RemoteUserAuthenticationBackend"]
AUTHENTICATION_HEADER = "AUTHORIZATION"
```

In docker nemo image I am receiving this error:
```
[16/Sep/2024 10:01:21] NEMO.middleware DEBUG Header: HTTP_AUTHORIZATION not present or invalid
[16/Sep/2024 10:01:21] NEMO.middleware DEBUG Header: HTTP_AUTHORIZATION not present or invalid
[16/Sep/2024 10:01:21] django.request WARNING Bad Request: /login/
Bad Request: /login/
```

Any idea what needs to be done?

**Reply by rptmat57** [reply]
My bad, you need to use `AUTHENTICATION_HEADER = "REMOTE_USER"` if you are using the shibboleth nginx image

Or change the line in Nginx `proxy_set_header Remote-User $shib_remote_user;` to `proxy_set_header Authorization $shib_remote_user;`

**Reply by Hydrozyk** [reply]
Thank you! 

I did as you mentioned but now getting this error:
```
nemo  | [16/Sep/2024 11:55:55] NEMO.middleware DEBUG Header: HTTP_HTTP_REMOTE_USER not present or invalid
nemo  | [16/Sep/2024 11:55:55] django.request WARNING Bad Request: /login/
nemo  | Bad Request: /login
```

I am using information for nginx configuration from: https://hub.docker.com/r/nanofab/nginx-shibboleth

I am wondering if nginx is not able to pass header information to NEMO? For example could it be that this directive:

`more_clear_input_headers 'Variable-*' 'Shib-*' 'Remote-User' 'REMOTE_USER' 'Auth-Type' 'AUTH_TYPE';`

Is clearing the REMOTE_USER header from the incoming request before it reaches the application. Therefore, even if I set REMOTE_USER later, it may not be passed correctly. Just thinking out loud here.

**Reply by rptmat57** [reply]
looks like you might be using "HTTP_REMOTE_USER" instead of "REMOTE_USER" in the `AUTHENTICATION_HEADER`
NEMO adds the HTTP_ automatically

**Reply by Hydrozyk** [reply]
It does not look like I was, here is my nginx location / config:

```
location / {
            include shib_fastcgi_params;
            include fastcgi_params;
             more_clear_input_headers 'Variable-*' 'Shib-*' 'Remote-User' 'REMOTE_USER' 'Auth-Type' 'AUTH_TYPE';
            fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
            shib_request_set $shib_remote_user $upstream_http_variable_remote_user;
            shib_request /shibauthorizer;
            proxy_http_version 1.1;
            proxy_set_header Host $host;
            proxy_set_header X-Forwarded-Host $server_name;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto https;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header Connection "";
            proxy_set_header REMOTE_USER $shib_remote_user;
            proxy_pass_request_headers on;
            proxy_pass http://172.18.0.3:8000;

        }
```

In Nginx logs I see that remote header is set to SHIB_REMOTE_USER
`16/Sep/2024:16:08:06 +0000] "GET /static/favicon.ico HTTP/1.1" 404 41676 "https:///" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/115.0" REMOTE_USER=- SHIB_REMOTE_USER=`

**Reply by rptmat57** [reply]
Sorry I was talking about the option in NEMO's `settings.py`

**Reply by Hydrozyk** [reply]
Yes corrected it but still same thing. I wonder if any of nginx settings need to change since, what it appears nemo is not receiving the REMOTE_USER header.

```
nemo  | [16/Sep/2024 17:00:43] NEMO.middleware DEBUG Header: HTTP_REMOTE_USER not present or invalid
nemo  | [16/Sep/2024 17:00:43] django.request WARNING Bad Request: /login/
nemo  | Bad Request: /login/
```

**Reply by Hydrozyk** [reply]
Also, should I open a separate issue thread in issues tab? let me know and I can do it.

**Reply by rptmat57** [reply]
not sure if it changes anything, but in my configuration I have `proxy_set_header Remote-User $shib_remote_user;` right before `shib_request /shibauthorizer;` and I don't have  `proxy_pass_request_headers on;` in my config at all.

in shibboleth2.xml, I have:
```
<ApplicationDefaults entityID="<entity_id>"
REMOTE_USER="uid eppn" 
cipherSuites="DEFAULT:!EXP:!LOW:!aNULL:!eNULL:!DES:!IDEA:!SEED:!RC4:!3DES:!kRSA:!SSLv2:!SSLv3:!TLSv1:!TLSv1.1">
```

and in my attribute-map.xml:
```
<Attribute name="username" id="uid"/>
```
